### ISSUE TITLE
default Timestamp behavior with Slick and MySQL

### ISSUE THREAD

**Issue by willgu-gdm** [slick/slick#2103]
Hi

I'm on slick version `3.2.3` and MySQL 5.7 (tried on 5.7.21/22/26 respectively).

I'm using the example provided in the official slick doc

```
case class User(id: Option[Int], first: String, last: String)

class Users(tag: Tag) extends Table[User](tag, "users") {
  def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
  def first = column[String]("first")
  def last = column[String]("last")
  def * = (id.?, first, last) <> (User.tupled, User.unapply)
}
val users = TableQuery[Users]
```

Here I'd change it to something similar by removing the `id` from the case class and manually code the projection

```
case class User(first: String, last: String)

class Users(tag: Tag) extends Table[User](tag, "users") {
  def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
  def first = column[String]("first")
  def last = column[String]("last")
  def * = (id.?, first, last) <> (
    { case (_, first, last) => User(first, last) },
    { u: Users => Some(None, u.first, u.last) }
  )
}
val users = TableQuery[Users]
```

and further I would like to do something like 

```
import java.sql.Timestamp

case class User(first: String, last: String)

class Users(tag: Tag) extends Table[User](tag, "users") {
  def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
  def first = column[String]("first")
  def last = column[String]("last")
  def createTime = column[Timestamp]("createTime", O.SqlType("timestamp(3) default current_timestamp(3)"))
  def * = (id.?, first, last, createTime.?) <> (
    { case (_, first, last, _) => User(first, last) },
    { u: Users => Some(None, u.first, u.last, None) }
  )
}
val users = TableQuery[Users]
```

but this does not necessarily work with MySQL. it depends on the option `explicit_defaults_for_timestamp`. if this option is `ON` then the data insertion leads to a run time error of `timestamp cannot be null`. I don't think slick should expect MySQL to prefer one side over the other, because sometimes people would rather keep the strict mode.

I think it's related to the way data insertion is done. take the following 2 SQL queries as example:
```
insert into users values (null, 'John', 'Doe', null); // does not work with strict mode on
insert into users(first, last) values ('John', 'Doe'); // works in both cases
```

so basically passing a `null` and expecting MySQL to set it to default doesn't always work.

**Comment by willgu-gdm** [comment]
after thinking this for a while, I figured that it's more of a MySQL thing. MySQL has no issue taking a `null` for `id` because `id` is the auto incremented primary key, so there's not ambiguity there. But for `timestamp not null` there might be, that why they don't like this behavior be default?

well, in this case, the following change would work
```
def createTime = column[Option[Timestamp]]("createTime", O.SqlType("timestamp(3) default current_timestamp(3)"))
```

now MySQL is happily taking a `null` for this column

**Comment by willgu-gdm** [comment]
sorry I have to re-open this issue because after I set the `createTime` column to `Option[Timestamp]`, all it shows are `NULL`s in MySQL. In this case, the default would not kick in.

So we are back to the start. I have to set MySQL `explicit_defaults_for_timestamp=0` to make this work.
